Exception when using source filtering on a large integer field

<!--

** Please read the guidelines below. **

Issues that do not follow these guidelines are likely to be closed.

1.  GitHub is reserved for bug reports and feature requests. The best place to
    ask a general question is at the Elastic [forums](https://discuss.elastic.co).
    GitHub is not the place for general questions.

2.  Is this bug report or feature request for a supported OS? If not, it
    is likely to be closed.  See https://www.elastic.co/support/matrix#show_os

3.  Please fill out EITHER the feature request block or the bug report block
    below, and delete the other block.

-->

<!-- Bug report -->

**Elasticsearch version** (`bin/elasticsearch --version`): 6.3.1 using the official docker image

**Plugins installed**: []

**Description of the problem including expected versus actual behavior**:

When indexing a document with a large integer field as (with `keyword` mappings`, querying fails with an error when using source filtering.
This seems a bit similar to #31853, but different enough that I thought I should open an issue.

**Steps to reproduce**:

 1. Using the official Elasticsearch package in python:
```python
es.indices.create('test-index')
es.indices.put_mapping(doc_type='doc', index='test-index', body={
    'properties': {
        'offset': {
            'type': 'keyword'
        }
    }
})
es.index('test-index', 'doc', {'offset': 2 ** 2500})
```

 2. While a regular query works as expected:
```python
>>> es.search(index='test-index', body={
    'query': {'match_all': {}}, 
    'size': 10
})

{ 'hits': { 'hits': [ '_source': {'offset': 37....} ...}
```

Using source-filtering will result in an exception:

```python
>>> es.search(index='test-index', body={
    '_source': ['offset'], 
    'query': {'match_all': {}}, 
    'size': 10
})

{'_shards': {'failed': 1,
  'failures': [{'index': 'test-index',
    'node': 'G2YpLXsGTNGrEphsO5SASQ',
    'reason': {'reason': 'cannot write xcontent for unknown value of type class java.math.BigInteger',
     'type': 'illegal_argument_exception'},
    'shard': 1}],
  'skipped': 0,
  'successful': 4,
  'total': 5},
 'hits': {'hits': [], 'max_score': 1.0, 'total': 1},
 'timed_out': False,
 'took': 8}
```

(Using `'_source': ['*']` will result in the same error)

**Provide logs (if relevant)**:
When using a **scroll** query, the server log shows the following error:
```bash
[2018-07-26T07:05:35,682][DEBUG][o.e.a.s.TransportSearchScrollAction] [server-name] [157387] Failed to execute fetch phase
org.elasticsearch.transport.RemoteTransportException: [server-name][10.0.0.7:9300][indices:data/read/search[phase/fetch/id/scroll]]
Caused by: java.lang.IllegalArgumentException: cannot write xcontent for unknown value of type class java.math.BigInteger
       at org.elasticsearch.common.xcontent.XContentBuilder.unknownValue(XContentBuilder.java:755) ~[elasticsearch-x-content-6.3.1.jar:6.3.1]
       at org.elasticsearch.common.xcontent.XContentBuilder.map(XContentBuilder.java:810) ~[elasticsearch-x-content-6.3.1.jar:6.3.1]
       at org.elasticsearch.common.xcontent.XContentBuilder.unknownValue(XContentBuilder.java:744) ~[elasticsearch-x-content-6.3.1.jar:6.3.1]
       at org.elasticsearch.common.xcontent.XContentBuilder.value(XContentBuilder.java:837) ~[elasticsearch-x-content-6.3.1.jar:6.3.1]
       at org.elasticsearch.common.xcontent.XContentBuilder.unknownValue(XContentBuilder.java:746) ~[elasticsearch-x-content-6.3.1.jar:6.3.1]
       at org.elasticsearch.common.xcontent.XContentBuilder.map(XContentBuilder.java:810) ~[elasticsearch-x-content-6.3.1.jar:6.3.1]
       at org.elasticsearch.common.xcontent.XContentBuilder.unknownValue(XContentBuilder.java:744) ~[elasticsearch-x-content-6.3.1.jar:6.3.1]
       at org.elasticsearch.common.xcontent.XContentBuilder.value(XContentBuilder.java:726) ~[elasticsearch-x-content-6.3.1.jar:6.3.1]
       at org.elasticsearch.search.fetch.subphase.FetchSourceSubPhase.hitExecute(FetchSourceSubPhase.java:65) ~[elasticsearch-6.3.1.jar:6.3.1]
       at org.elasticsearch.search.fetch.FetchPhase.execute(FetchPhase.java:162) ~[elasticsearch-6.3.1.jar:6.3.1]
       at org.elasticsearch.search.SearchService.executeFetchPhase(SearchService.java:516) ~[elasticsearch-6.3.1.jar:6.3.1]
       at org.elasticsearch.action.search.SearchTransportService$10.messageReceived(SearchTransportService.java:430) ~[elasticsearch-6.3.1.jar:6.3.1]
       at org.elasticsearch.action.search.SearchTransportService$10.messageReceived(SearchTransportService.java:427) ~[elasticsearch-6.3.1.jar:6.3.1]
       at org.elasticsearch.transport.RequestHandlerRegistry.processMessageReceived(RequestHandlerRegistry.java:66) ~[elasticsearch-6.3.1.jar:6.3.1]
       at org.elasticsearch.transport.TransportService$7.doRun(TransportService.java:664) [elasticsearch-6.3.1.jar:6.3.1]
       at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:725) [elasticsearch-6.3.1.jar:6.3.1]
       at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37) [elasticsearch-6.3.1.jar:6.3.1]
       at org.elasticsearch.common.util.concurrent.TimedRunnable.doRun(TimedRunnable.java:41) [elasticsearch-6.3.1.jar:6.3.1]
       at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37) [elasticsearch-6.3.1.jar:6.3.1]
       at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1135) [?:?]
       at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) [?:?]
       at java.lang.Thread.run(Thread.java:844) [?:?]
```

I wonder if we should even allow clients to handle `BigInteger` or `BigDecimal` fields. 
We don't allow indexing big integers through http layer.
The following attempt to index to the keyword field a number > Long.MAX_VALUE produces an error.

```js
PUT test1
{
  "mappings": {
    "_doc": {
      "properties": {
        "text1" : "keyword"
      }
    }
  }
}

PUT test1/_doc/1
{
  "test1" :  9223372036854775808
}
```

```
{
	"error": {
		"root_cause": [
			{
				"type": "mapper_parsing_exception",
				"reason": "failed to parse field [test1] of type [long]"
			}
		],
		"type": "mapper_parsing_exception",
		"reason": "failed to parse field [test1] of type [long]",
		"caused_by": {
			"type": "json_parse_exception",
			"reason": "Numeric value (9223372036854775808) out of range of long (-9223372036854775808 - 9223372036854775807)\n at [Source: org.elasticsearch.common.bytes.BytesReference$MarkSupportingStreamInputWrapper@1fd04cca; line: 2, column: 33]"
		}
	},
	"status": 400
}

[elasticsearch] Caused by: com.fasterxml.jackson.core.JsonParseException: Numeric value (922337203685477580792233720368547758079223372036854775807) out of range of long (-9223372036854775808 - 9223372036854775807)
[elasticsearch]  at [Source: org.elasticsearch.common.bytes.BytesReference$MarkSupportingStreamInputWrapper@14afdda; line: 2, column: 71]
```

--- a/libs/x-content/src/main/java/org/elasticsearch/common/xcontent/XContentBuilder.java
+++ b/libs/x-content/src/main/java/org/elasticsearch/common/xcontent/XContentBuilder.java
@@ -25,6 +25,8 @@ import java.io.Flushable;
 import java.io.IOException;
 import java.io.InputStream;
 import java.io.OutputStream;
+import java.math.BigDecimal;
+import java.math.BigInteger;
 import java.nio.file.Path;
 import java.time.ZonedDateTime;
 import java.util.Arrays;
@@ -103,7 +105,8 @@ public final class XContentBuilder implements Closeable, Flushable {
         writers.put(ZonedDateTime.class, (b, v) -> b.value(v.toString()));
         writers.put(Calendar.class, XContentBuilder::timeValue);
         writers.put(GregorianCalendar.class, XContentBuilder::timeValue);
-
+        writers.put(BigInteger.class, (b, v) -> b.value((BigInteger) v));
+        writers.put(BigDecimal.class, (b, v) -> b.value((BigDecimal) v));
 
         Map<Class<?>, HumanReadableTransformer> humanReadableTransformer = new HashMap<>();
         Map<Class<?>, Function<Object, Object>> dateTransformers = new HashMap<>();
@@ -542,6 +545,81 @@ public final class XContentBuilder implements Closeable, Flushable {
     }
 
     public XContentBuilder value(short value) throws IOException {
+        generator.writeNumber(value);
+        return this;
+    }
+
+    ////////////////////////////////////////////////////////////////////////////
+    // BigInteger
+    //////////////////////////////////
+
+    public XContentBuilder field(String name, BigInteger value) throws IOException {
+        if (value == null) {
+            return nullField(name);
+        }
+        ensureNameNotNull(name);
+        generator.writeNumberField(name, value);
+        return this;
+    }
+
+    public XContentBuilder array(String name, BigInteger[] values) throws IOException {
+        return field(name).values(values);
+    }
+
+    private XContentBuilder values(BigInteger[] values) throws IOException {
+        if (values == null) {
+            return nullValue();
+        }
+        startArray();
+        for (BigInteger b : values) {
+            value(b);
+        }
+        endArray();
+        return this;
+    }
+
+    public XContentBuilder value(BigInteger value) throws IOException {
+        if (value == null) {
+            return nullValue();
+        }
+        generator.writeNumber(value);
+        return this;
+    }
+
+
+    ////////////////////////////////////////////////////////////////////////////
+    // BigDecimal
+    //////////////////////////////////
+
+    public XContentBuilder field(String name, BigDecimal value) throws IOException {
+        if (value == null) {
+            return nullField(name);
+        }
+        ensureNameNotNull(name);
+        generator.writeNumberField(name, value);
+        return this;
+    }
+
+    public XContentBuilder array(String name, BigDecimal[] values) throws IOException {
+        return field(name).values(values);
+    }
+
+    private XContentBuilder values(BigDecimal[] values) throws IOException {
+        if (values == null) {
+            return nullValue();
+        }
+        startArray();
+        for (BigDecimal b : values) {
+            value(b);
+        }
+        endArray();
+        return this;
+    }
+
+    public XContentBuilder value(BigDecimal value) throws IOException {
+        if (value == null) {
+            return nullValue();
+        }
         generator.writeNumber(value);
         return this;
     }

--- a/libs/x-content/src/main/java/org/elasticsearch/common/xcontent/XContentGenerator.java
+++ b/libs/x-content/src/main/java/org/elasticsearch/common/xcontent/XContentGenerator.java
@@ -23,6 +23,8 @@ import java.io.Closeable;
 import java.io.Flushable;
 import java.io.IOException;
 import java.io.InputStream;
+import java.math.BigDecimal;
+import java.math.BigInteger;
 
 public interface XContentGenerator extends Closeable, Flushable {
 
@@ -69,6 +71,14 @@ public interface XContentGenerator extends Closeable, Flushable {
     void writeNumber(long value) throws IOException;
 
     void writeNumber(short value) throws IOException;
+
+    void writeNumber(BigInteger value) throws IOException;
+
+    void writeNumberField(String name, BigInteger value) throws IOException;
+
+    void writeNumber(BigDecimal value) throws IOException;
+
+    void writeNumberField(String name, BigDecimal value) throws IOException;
 
     void writeStringField(String name, String value) throws IOException;
 

--- a/libs/x-content/src/main/java/org/elasticsearch/common/xcontent/json/JsonXContentGenerator.java
+++ b/libs/x-content/src/main/java/org/elasticsearch/common/xcontent/json/JsonXContentGenerator.java
@@ -42,6 +42,8 @@ import java.io.BufferedInputStream;
 import java.io.IOException;
 import java.io.InputStream;
 import java.io.OutputStream;
+import java.math.BigDecimal;
+import java.math.BigInteger;
 import java.util.Objects;
 import java.util.Set;
 
@@ -227,6 +229,19 @@ public class JsonXContentGenerator implements XContentGenerator {
     }
 
     @Override
+    public void writeNumberField(String name, BigInteger value) throws IOException {
+        // as jackson's JsonGenerator doesn't have this method for BigInteger
+        // we have to implement it ourselves
+        generator.writeFieldName(name);
+        generator.writeNumber(value);
+    }
+
+    @Override
+    public void writeNumberField(String name, BigDecimal value) throws IOException {
+        generator.writeNumberField(name, value);
+    }
+
+    @Override
     public void writeNumber(int value) throws IOException {
         generator.writeNumber(value);
     }
@@ -243,6 +258,16 @@ public class JsonXContentGenerator implements XContentGenerator {
 
     @Override
     public void writeNumber(short value) throws IOException {
+        generator.writeNumber(value);
+    }
+
+    @Override
+    public void writeNumber(BigInteger value) throws IOException {
+        generator.writeNumber(value);
+    }
+
+    @Override
+    public void writeNumber(BigDecimal value) throws IOException {
         generator.writeNumber(value);
     }
 

--- a/rest-api-spec/src/main/resources/rest-api-spec/test/search/10_source_filtering.yml
+++ b/rest-api-spec/src/main/resources/rest-api-spec/test/search/10_source_filtering.yml
@@ -112,6 +112,16 @@ setup:
   - match:  { hits.hits.0._source.bigint: 72057594037927936 }
   - is_false:  hits.hits.0._source.include.field2
 
+
+---
+"_source filtering on bigint":
+- do:
+    search:
+      body:
+        _source: ["bigint"]
+        query: { match_all: {} }
+- match:  { hits.hits.0._source.bigint: 72057594037927936 }
+
 ---
 "fields in body":
   - do:

--- a/server/src/test/java/org/elasticsearch/common/xcontent/BaseXContentTestCase.java
+++ b/server/src/test/java/org/elasticsearch/common/xcontent/BaseXContentTestCase.java
@@ -48,6 +48,7 @@ import org.joda.time.format.ISODateTimeFormat;
 import java.io.ByteArrayInputStream;
 import java.io.ByteArrayOutputStream;
 import java.io.IOException;
+import java.math.BigDecimal;
 import java.math.BigInteger;
 import java.nio.file.Path;
 import java.time.DayOfWeek;
@@ -264,6 +265,36 @@ public abstract class BaseXContentTestCase extends ESTestCase {
                 .startObject()
                     .array("short", new short[]{Short.MAX_VALUE, Short.MIN_VALUE})
                 .endObject());
+    }
+
+    public void testBigIntegers() throws Exception {
+        assertResult("{'bigint':null}", () -> builder().startObject().field("bigint", (BigInteger) null).endObject());
+        assertResult("{'bigint':[]}", () -> builder().startObject().array("bigint", new BigInteger[]{}).endObject());
+
+        BigInteger bigInteger = BigInteger.valueOf(Long.MAX_VALUE).add(BigInteger.ONE);
+        String result = "{'bigint':" + bigInteger.toString() + "}";
+        assertResult(result, () -> builder().startObject().field("bigint", bigInteger).endObject());
+
+        result = "{'bigint':[" + bigInteger.toString() + "," + bigInteger.toString() + "," + bigInteger.toString() +"]}";
+        assertResult(result, () -> builder()
+            .startObject()
+            .array("bigint", bigInteger, bigInteger, bigInteger)
+            .endObject());
+    }
+
+    public void testBigDecimals() throws Exception {
+        assertResult("{'bigdecimal':null}", () -> builder().startObject().field("bigdecimal", (BigInteger) null).endObject());
+        assertResult("{'bigdecimal':[]}", () -> builder().startObject().array("bigdecimal", new BigInteger[]{}).endObject());
+
+        BigDecimal bigDecimal = new BigDecimal("234.43");
+        String result = "{'bigdecimal':" + bigDecimal.toString() + "}";
+        assertResult(result, () -> builder().startObject().field("bigdecimal", bigDecimal).endObject());
+
+        result = "{'bigdecimal':[" + bigDecimal.toString() + "," + bigDecimal.toString() + "," + bigDecimal.toString() +"]}";
+        assertResult(result, () -> builder()
+            .startObject()
+            .array("bigdecimal", bigDecimal, bigDecimal, bigDecimal)
+            .endObject());
     }
 
     public void testStrings() throws IOException {

